src/messaging/ReliableMessageManager.cpp

```diff
@@ -0,0 +1,656 @@
+        err = CHIP_ERROR_RETRANS_TABLE_FULL;
+    }
+
+    return err;
+}
+
+void ReliableMessageManager::PauseRetransTable(ReliableMessageContext * rc, uint32_t PauseTimeMillis)
+{
+    for (int i = 0; i < CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE; i++)
+    {
+        if (RetransTable[i].rc == rc)
+        {
+            RetransTable[i].nextRetransTimeTick =
+                static_cast<uint16_t>(RetransTable[i].nextRetransTimeTick + (PauseTimeMillis >> mTimerIntervalShift));
+            break;
+        }
+    }
+}
+
+void ReliableMessageManager::ResumeRetransTable(ReliableMessageContext * rc)
+{
+    for (int i = 0; i < CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE; i++)
+    {
+        if (RetransTable[i].rc == rc)
+        {
+            RetransTable[i].nextRetransTimeTick = 0;
+            break;
+        }
+    }
+}
+
+bool ReliableMessageManager::CheckAndRemRetransTable(ReliableMessageContext * rc, uint32_t ackMsgId)
+{
+    bool res = false;
+
+    for (int i = 0; i < CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE; i++)
+    {
+        if ((RetransTable[i].rc == rc) && RetransTable[i].msgId == ackMsgId)
+        {
+            // Clear the entry from the retransmision table.
+            ClearRetransmitTable(RetransTable[i]);
+
+#if defined(DEBUG)
+            ChipLogProgress(ExchangeManager, "Rxd Ack; Removing MsgId:%08" PRIX32 " from Retrans Table", ackMsgId);
+#endif
+            res = true;
+            break;
+        }
+
+        continue;
+    }
+
+    return res;
+}
+
+/**
+ *  Send the specified entry from the retransmission table.
+ *
+ *  @param[in]    entry                A pointer to a retransmission table entry object that needs to be sent.
+ *
+ *  @return  #CHIP_NO_ERROR On success, else corresponding CHIP_ERROR returned from SendMessage.
+ *
+ */
+CHIP_ERROR ReliableMessageManager::SendFromRetransTable(RetransTableEntry * entry)
+{
+    CHIP_ERROR err              = CHIP_NO_ERROR;
+    ReliableMessageContext * rc = entry->rc;
+
+    // To trigger a call to OnSendError, set the number of transmissions so
+    // that the next call to ExecuteActions will abort this entry,
+    // restart the timer immediately, and ExitNow.
+
+    CHIP_FAULT_INJECT(FaultInjection::kFault_RMPSendError, entry->sendCount = static_cast<uint8_t>(rc->mConfig.mMaxRetrans + 1);
+                      entry->nextRetransTimeTick = 0; StartTimer(); ExitNow());
+
+    if (rc)
+    {
+        // Locally store the start and length;
+        uint8_t * p  = entry->msgBuf->Start();
+        uint32_t len = entry->msgBuf->DataLength();
+
+        // Send the message through
+        uint16_t msgSendFlags = entry->msgSendFlags;
+        SetFlag(msgSendFlags, kChipMessageFlag_RetainBuffer);
+        err = rc->mDelegate.SendMessage(entry->msgBuf, msgSendFlags);
+
+        // Reset the msgBuf start pointer and data length after sending
+        entry->msgBuf->SetStart(p);
+        entry->msgBuf->SetDataLength(static_cast<uint16_t>(len));
```

Comment:
could we make len be uint16_t instead and not need to cast? It seems as if Get and Set should support he same data size.

Reply:
done